packages/backend/test/utils.ts

```diff
@@ -364,3 +364,68 @@ export const userStateTransition = async (t, iden) => {
     await waitForBackendBlock(t, receipt.blockNumber)
     t.pass()
 }
+
+const genUsernameProof = async (t, iden, graffitiPreImage) => {
+    const userState = await genUserState(
+        t.context.unirepSocial.provider,
+        t.context.unirep.address,
+        iden
+    )
+    // find valid nonce starter
+    // gen proof
+
+    const epkNonce = 0
+    const usernameProof = await userState.genProveReputationProof(
+        t.context.attesterId,
+        epkNonce,
+        0,
+        BigInt(0),
+        graffitiPreImage,
+        BigInt(0)
+    )
+
+    const isValid = await usernameProof.verify()
+    t.true(isValid)
+
+    // we need to wait for the backend to process whatever block our provider is on
+    const blockNumber = await t.context.provider.getBlockNumber()
+    return {
+        proof: usernameProof.proof,
+        publicSignals: usernameProof.publicSignals,
+        blockNumber,
+    }
+}
+
+export const setUsername = async (t, iden) => {
+    const preImage = genRandomSalt().valueOf()
+    const { proof, publicSignals, blockNumber } = await genUsernameProof(
+        t,
+        iden,
+        preImage
+    )
+    await waitForBackendBlock(t, blockNumber)
+
+    const r = await fetch(`${t.context.url}/api/usernames`, {
+        method: 'POST',
+        headers: {
+            'content-type': 'application/json',
+        },
+        body: JSON.stringify({
+            newUsername: 15,
```

Comment:
Try making this a proper string. It might be good to apply some username restrictions in the backend, numbers and characters plus some other characters.

Reply:
So would we want to accept some username like `'testusername-123'` and then hash this string with poseidon `hashOne`? Then the string username has to be converted to bigint type before right?

Reply:
Yes. We need to decide how we're going to convert text to bytes. I think it's okay to utf8 with 32 bytes. This will allow a username to be up to 32 characters. Then there it should be pretty easy to convert to hex and use it as an input to poseidon.

Once you have a hex string you can make a bigint using `BigInt(hexString)`.

We also shouldn't allow some characters, like spaces probably.

Reply:
You can use ethers to convert a string to utf8

https://docs.ethers.io/v5/api/utils/strings/#strings-utf8